tensorflow/c/experimental/gradients/math_grad_test.cc

```diff
@@ -0,0 +1,365 @@
+#include "tensorflow/c/tf_status_helper.h"
+#include "tensorflow/core/platform/tensor_float_32_utils.h"
+#include "tensorflow/core/platform/test.h"
+
+namespace tensorflow {
+namespace gradients {
+namespace internal {
+namespace {
+
+using tensorflow::TF_StatusPtr;
+
+Status AddModel(AbstractContext* ctx,
+                absl::Span<AbstractTensorHandle* const> inputs,
+                absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Add(ctx, inputs, outputs, "Add");
+}
+
+Status ExpModel(AbstractContext* ctx,
+                absl::Span<AbstractTensorHandle* const> inputs,
+                absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Exp(ctx, inputs, outputs, "Exp");
+}
+
+Status SqrtModel(AbstractContext* ctx,
+                 absl::Span<AbstractTensorHandle* const> inputs,
+                 absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Sqrt(ctx, inputs, outputs, "Sqrt");
+}
+
+Status NegModel(AbstractContext* ctx,
+                absl::Span<AbstractTensorHandle* const> inputs,
+                absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Neg(ctx, inputs, outputs, "Neg");
+}
+
+Status SubModel(AbstractContext* ctx,
+                absl::Span<AbstractTensorHandle* const> inputs,
+                absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Sub(ctx, inputs, outputs, "Sub");
+}
+
+Status MulModel(AbstractContext* ctx,
+                absl::Span<AbstractTensorHandle* const> inputs,
+                absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Mul(ctx, inputs, outputs, "Mul");
+}
+
+Status Log1pModel(AbstractContext* ctx,
+                  absl::Span<AbstractTensorHandle* const> inputs,
+                  absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::Log1p(ctx, inputs, outputs, "Log1p");
+}
+
+Status DivNoNanModel(AbstractContext* ctx,
+                     absl::Span<AbstractTensorHandle* const> inputs,
+                     absl::Span<AbstractTensorHandle*> outputs) {
+  return ops::DivNoNan(ctx, inputs, outputs, "DivNoNan");
+}
+
+class CppGradients
+    : public ::testing::TestWithParam<std::tuple<const char*, bool, bool>> {
+ protected:
+  void SetUp() override {
+    TF_StatusPtr status(TF_NewStatus());
+    TF_SetTracingImplementation(std::get<0>(GetParam()), status.get());
+    status_ = StatusFromTF_Status(status.get());
+    ASSERT_EQ(errors::OK, status_.code()) << status_.error_message();
+
+    {
+      AbstractContext* ctx_raw = nullptr;
+      Status s =
+          BuildImmediateExecutionContext(std::get<1>(GetParam()), &ctx_raw);
+      ASSERT_EQ(errors::OK, s.code()) << s.error_message();
+      ctx_.reset(ctx_raw);
+    }
+
+    // Computing numerical gradients with TensorFloat-32 is numerically
+    // unstable. Some forward pass tests also fail with TensorFloat-32 due to
+    // low tolerances
+    enable_tensor_float_32_execution(false);
+  }
+
+  AbstractContextPtr ctx_;
+  GradientRegistry registry_;
```

Comment:
Do we need to declare `registry_` and `status_` here or can these be local to the tests?

Reply:
they are indeed local to the tests. You could see that in `When these tests run, the following happens:` in https://github.com/google/googletest/blob/master/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests

Reply:
Yes, but I was wondering if we gain much from declaring these here?

Reply:
I think it saves us some lines ( with `registry_` ) or allows us to reuse the variable ( not really but with the previous code, we just created a `status`, used it once and threw it away. I think it is not really the best way ). WDYT ?